[HUDI-4543] Support natural order when table schema contains a field named 'ts'

flashJd opened the pull request:
## What is the purpose of the pull request
when migrate external data  to hudi, we extract the schema of the source table and use it to create hudi table.
but when the schema contains a column named ts(a common situation), hudi will always precombine data using this column and we can't disable it.

## Brief change log
Be able to disable precombine field when table schema contains a field named ts

## Verify this pull request
modify a test: testRequiredOptionsForSource in TestHoodieTableFactory.java
add a test: testWriteAndReadWithProctimeSequenceWithTsColumnExisting in ITTestHoodieDataSource.java

flashJd commented:
@dannyhchen can you review it

flashJd commented:
@danny0405 can you review it

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/HoodieTableFactory.java:
```diff
@@ -147,7 +147,7 @@ private void sanityCheck(Configuration conf, ResolvedSchema schema) {
       }
-      if (preCombineField.equals(FlinkOptions.PRECOMBINE_FIELD.defaultValue())) {
+      if (preCombineField.equals(FlinkOptions.PRECOMBINE_FIELD.defaultValue()) || preCombineField.equals(FlinkOptions.NO_PRE_COMBINE)){
         conf.setString(FlinkOptions.PRECOMBINE_FIELD, FlinkOptions.NO_PRE_COMBINE);
```
Is this change valid ? We still set up the option if it is already `FlinkOptions.NO_PRE_COMBINE`.

flashJd replied:
it's valid, suppose a table with a 'ts' column, we want to set PRECOMBINE_FIELD=FlinkOptions.NO_PRE_COMBINE to disable the precombine field, above code will go into line 153, because preCombineField.equals(FlinkOptions.PRECOMBINE_FIELD.defaultValue()) is false, FlinkOptions.PRECOMBINE_FIELD.defaultValue() is 'ts', preCombineField is what we set 'no_precombine', you can also see the test.

flashJd replied:
> Is this change valid ? We still set up the option if it is already `FlinkOptions.NO_PRE_COMBINE`.

@danny0405 looking forward to your reply

danny0405 replied:
One thing need to note here is that: option `FlinkOptions.NO_PRE_COMBINE` is only for internal use

flashJd replied:
yes, i want the proc_time/natural order sequence semantics even though you got a ts field in the schema

danny0405 replied:
Then we can do such logic:
else if (!preCombineField.equals(FlinkOptions.NO_PRE_COMBINE)) {
    throw new HoodieValidationException("Field " + preCombineField + " does not exist in the table schema."
            + "Please check '" + FlinkOptions.PRECOMBINE_FIELD.key() + "' option.");
}

flashJd replied:
yes, it is more elegant. I've modified to it

danny0405 commented:
Thanks for the contribution, can we log a JIRA issue and change the commit titile to form like:
`[HUDI-${JIRA issue ID}] ${your actual commit title}`.

flashJd commented:
> 
I've filed a Jira ticket and changed the commit titile, also fixed the checkStyle conflicts, can you help approval the workflow, thanks.

flashJd commented:
> Thanks for the contribution, can we log a JIRA issue and change the commit titile to form like: `[HUDI-${JIRA issue ID}] ${your actual commit title}`.

I've filed a Jira ticket and changed the commit titile, also fixed the checkStyle conflicts, can you help approval the workflow, thanks.

flashJd commented:
> > Thanks for the contribution, can we log a JIRA issue and change the commit titile to form like: `[HUDI-${JIRA issue ID}] ${your actual commit title}`.
> 
> I've filed a Jira ticket and changed the commit titile, also fixed the checkStyle conflicts, can you help approval the workflow, thanks.

@danny0405 can we merge it to the master
